src/hotspot/share/opto/memnode.cpp

```diff
@@ -3519,17 +3519,130 @@ Node *StoreNode::Ideal_masked_input(PhaseGVN *phase, uint mask) {
+// array_of_shorts[0] = (short)(X << 2)
+// We get a structure such as:
+// (StoreB ... (RShiftI _ (LShiftI _ (LShiftI _ X 2) 16) 16))
+// that is simplified into
+// (StoreB ... (RShiftI _ (LShiftI _ X 18) 16)).
+// It is thus useful to handle the case where conIL > conIR.
+//
+// Let's assume we have the following 32 bits integer that we want to stuff in 8 bits char:
+// +------------------------+---------+
+// |        v[8..31]        | v[0..7] |
+// +------------------------+---------+
+//  31                     8 7        0
+// v[0..7] is meaningful, but v[8..31] is not. In this case, num_rejected_bits == 24.
+// Let's study what happens in different cases to see that the simplification into
+// (StoreB ... (LShiftI _ valIn (conIL - conIR)) )
+// is valid if:
+// - conIL >= conIR
+// - conIR <= num_rejected_bits
+// Let's also remember that conIL < 32 since (x << 33) is simplified into (x << 1)
+// and (x << 31) << 2 is simplified into 0. This means that in any case, after the
+// left shift, we always have at least one bit of the original v.
+//
+// ### Case 1 : conIL == conIR
+// If we do the shift left then right by 24 bits, we get:
+// after << 24
+// +---------+------------------------+
+// | v[0..7] |           0            |
+// +---------+------------------------+
+//  31     24 23                      0
+// after >> 24
+// +------------------------+---------+
+// |        sign bit        | v[0..7] |
+// +------------------------+---------+
+//  31                     8 7        0
+// The non-rejected bits are the same before and after, so, indeed, simplifying is fine.
+//
+// ### Case 2: conIL > conIR == num_rejected_bits
+// We take conIL == 26 for this example.
+// after << 26
+// +---------+------------------------+
+// | v[0..5] |           0            |
+// +---------+------------------------+
+//  31     26 25                      0
+// after >> 24
+// +------------------+---------+-----+
+// |     sign bit     | v[0..5] |  0  |
+// +------------------+---------+-----+
+//  31               8 7       2 1   0
+// The non-rejected bits are the 8 lower one of (v << conIL - conIR).
+// The bits 6 and 7 of v have been thrown away after the shift left.
+// The simplification is still fine.
+//
+// ### Case 3: conIL > conIR < num_rejected_bits.
+// Let's say conIL == 26 and conIR == 22.
+// after << 26
+// +---------+------------------------+
+// | v[0..5] |           0            |
+// +---------+------------------------+
+//  31     26 25                      0
+// after >> 22
+// +------------------+---------+-----+
+// |     sign bit     | v[0..5] |  0  |
+// +------------------+---------+-----+
+//  31              10 9       4 3   0
+// The non-rejected bits are the 8 lower one of (v << conIL - conIR).
+// The bits 6 and 7 of v have been thrown away after the shift left.
+// The bits 4 and 5 of v are still present, but outside of the kept bits (the 8 lower ones).
```

Comment:
I have trouble understanding this line. Bits 0-5 are still present. What do you mean by "outside of the kept bits"?